src/components/SettingComponent.js

```diff
@@ -0,0 +1,215 @@
+/* eslint-disable react/jsx-props-no-spreading */
+/* eslint-disable object-curly-newline */
+import React, { useContext, useState } from 'react'
+import { string, func, any, arrayOf, number, bool } from 'prop-types'
+
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faEyeDropper } from '@fortawesome/free-solid-svg-icons'
+import classNames from 'classnames'
+
+import {
+  Switch,
+  Select,
+  MenuItem,
+  Slider as MaterialSlider,
+  Button as MaterialButton,
+  Popover,
+} from '@material-ui/core'
+
+import { SketchPicker } from 'react-color'
+
+import { OPTION_TYPES } from '../lib/options'
+import { SettingsContext } from '../lib/contexts'
+
+const generalPropTypes = {
+  option: string.isRequired,
+  onChange: func,
+}
+
+const generalDefaultProps = {
+  onChange: () => {},
+}
+
+const GeneralSettingEvent = Component => {
+  const HOC = ( { option, onChange, ...props } ) => (
+    <Component {...props} onChange={( { target: { value } } ) => onChange( option, value )} />
+  )
+
+  HOC.propTypes = generalPropTypes
+  HOC.defaultProps = generalDefaultProps
+
+  return HOC
+}
+
+const GeneralSettingParam = Component => {
+  const HOC = ( { option, onChange, ...props } ) => (
+    <Component {...props} onChange={( _, value ) => onChange( option, value )} />
+  )
+
+  HOC.propTypes = generalPropTypes
+  HOC.defaultProps = generalDefaultProps
+
+  return HOC
+}
+
+export const Toggle = ( { value, ...props } ) => <Switch className={classNames( 'toggle', { checked: value } )} checked={value} {...props} />
+
+Toggle.propTypes = {
+  value: bool.isRequired,
+}
+
+export const Slider = ( { name, min, max, step, storageKey, units, ...props } ) => {
+  const [ settings, setSettings ] = useContext( SettingsContext )
+
+  return (
+    <MaterialSlider
+      className="slider"
+      valueLabelDisplay="auto"
+      value={settings[name] || window.localStorage.getItem( storageKey ).split( units )[0]}
```

Comment:
I guess it's ok for now, but do be very aware that if `settings[name]` is false, then every re-render will cause a bunch of loads from localstorage (which is blocking and slower than reading an in-app object)